File: leetcode/107.binary-tree-level-order-traversal-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        
        if (root == NULL){
            return vector<vector<int>>();
        }
        
        vector<vector<int>> res;
        queue<TreeNode *> q;
        q.push(root);
        
        while (!q.empty()) {
            int size = q.size();
            vector<int> ans;
            for (auto i = 0; i < size; ++i)
            {
                TreeNode *node = q.front();
                q.pop();
                ans.push_back(node->val);
                
                if (node->left)
                {
                    q.push(node->left);
                }
                
                if (node->right)
                {
                    q.push(node->right);
                }
            }
            res.push_back(ans);
        }
        
        reverse(res.begin(), res.end());
        
        return res;
    }
};

File: leetcode/107.binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        
        res = []
        q = [root]
        
        while len(q) > 0:
            size = len(q)
            ans = []
            for i in range(size):
                tmp = q[0]
                q.pop(0)
                
                ans.append(tmp.val)
                
                if tmp.left is not None:
                    q.append(tmp.left)
                    
                if tmp.right is not None:
                    q.append(tmp.right)
                    
            res.append(ans)
            
        return res[::-1]